### PR TITLE
feature: improves Hash repos to match Document repos, fully tests EntityStreams, general clean up

### PR DESCRIPTION
**Pull request by bsbodden**
First, my apologies for the massive PR. This started as an effort to up the test coverage to 75% or higher. Along the way, I found a few issues (see below) that got fixed and a few that were labeled as "features" since they required new classes. 

[1]  8f0b2ef - (HEAD -> bsb/clean_up_and_refactoring, origin/bsb/clean_up_and_refactoring, main) test: add tests for modules operations API 
[2]  19bc74b - feature: finish implementing terminal JSON actions on EntityStreams 
[3]  f3cafc5 - refactor: consolidate ignore exceptions in BloomQueryExecutor 
[4]  a8de734 - refactor: remove sysouts and format 
[5]  f76178b - test: add test for language stemming in queries 
[6]  37186c2 - refactor: remove unused template from test, clear all repos before tests 
[7]  007f5f7 - test: cover all cases of Ulid ID generation 
[8]  6a7bc9c - feature: add getAllKeysOf method to retrieve paginated keys 
[9]  d7a7788 - refactor: add pass-thru RedisOMCustomConversions to RedisJSONKeyValueAdapter 
[10]  8ae1f73 - refactor: consolidate get methods 
[11]  f13213d - feature: introduce keyspace resolution for document repositories 
[12]  837d28e - test: add test when deleting all suggestion entries 
[13]  226505c - feature: handle metamodel generation for Hashed mapped fields, and general clean up of the metamodel API 
[14]  6a619fe - feature: improve Hash repositories to catch up with Document repositories querying features 
[15]  ba6a7f4 - feature: add missing Gson adapters for Ulid to GsonBuilderFactor 
[16]  0279a74 - test: add tests for ObjectUtils class 
[17]  aeea951 - refactor: clean query utils and account for custom conversions (for hashes) 
[18]  1de6cca - feature: add RedisOMCustomConversions to handle type conversions for Hash mappings 
[19]  e802a18 - refactor: correct invalid identifier type message 
[20]  5f9c040 - refactor: remove printStackTrace calls, default empty return value on exception 
[21]  1091d61 - feature: introduce KeyspaceToIndexMap to track Entity Indexes 
[22]  c68d4db - refactor: test and clean tuples implementation 
[23]  d01ff3d - refactor: add accessor for underlying RedisKeyValueAdapter in CustomRedisKeyValueTemplate 
[24]  38b3def - refactor: replace printStackTrace calls with logging statements 
[25]  e5f3ce9 - fix: add missing apply (aggregation transformations) parameter

### CODE REVIEW

File: redis-om-spring/src/main/java/com/redis/om/spring/convert/BooleanToBytesConverter.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package com.redis.om.spring.convert;
```
Comment (chayim) [inline redis-om-spring/src/main/java/com/redis/om/spring/convert/BooleanToBytesConverter.java:1]: If this level of conversation is problematic, to the point that we have to write multiple conversions - IMHO they all belong in Jedis itself.  This is generally the pattern relied upon in the python ecosystem.

File: redis-om-spring/src/main/java/com/redis/om/spring/convert/BooleanToBytesConverter.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+import org.springframework.core.convert.converter.Converter;
+import org.springframework.data.convert.WritingConverter;
+import org.springframework.stereotype.Component;
+
+@Component
+@WritingConverter
+public class BooleanToBytesConverter implements Converter<Boolean, byte[]> {
+  byte[] fromString(String source) {
+    return source.getBytes(StandardCharsets.UTF_8);
```
Comment (chayim) [inline redis-om-spring/src/main/java/com/redis/om/spring/convert/BooleanToBytesConverter.java:13]: How can you be sure they're UTF_8?

File: redis-om-spring/src/main/java/com/redis/om/spring/convert/BytesToBooleanConverter.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    if (ObjectUtils.isEmpty(source)) {
+      return null;
+    }
+
+    String value = toString(source);
+    return ("1".equals(value) || "true".equalsIgnoreCase(value)) ? Boolean.TRUE : Boolean.FALSE;
+  }
+
+  String toString(byte[] source) {
+    return new String(source, StandardCharsets.UTF_8);
```
Comment (chayim) [inline redis-om-spring/src/main/java/com/redis/om/spring/convert/BytesToBooleanConverter.java:23]: UTF_8 - but we should support multiple encodings ideally. Clients generally do.

File: redis-om-spring/src/main/java/com/redis/om/spring/convert/BytesToUlidConverter.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
+package com.redis.om.spring.convert;
+
+import java.nio.charset.StandardCharsets;
+
+import org.springframework.core.convert.converter.Converter;
+import org.springframework.data.convert.ReadingConverter;
+import org.springframework.util.ObjectUtils;
+
+import com.github.f4b6a3.ulid.Ulid;
```
Comment (chayim) [inline redis-om-spring/src/main/java/com/redis/om/spring/convert/BytesToUlidConverter.java:9]: @bsbodden pardon my ignorance here! I found a lot of use of [sulky ulid](https://mvnrepository.com/artifact/de.huxhorn.sulky/de.huxhorn.sulky.ulid) all over RedHat, Netflix, and others. Is one better than the other?

Not that it's worth anything, but that library is also far higher up on maven.org

File: redis-om-spring/src/main/java/com/redis/om/spring/metamodel/indexed/DateField.java [hunk 5]
```diff
@@ -4,17 +4,27 @@
 
 import com.redis.om.spring.metamodel.MetamodelField;
 import com.redis.om.spring.search.stream.predicates.numeric.BetweenPredicate;
+import com.redis.om.spring.search.stream.predicates.numeric.EqualPredicate;
 import com.redis.om.spring.search.stream.predicates.numeric.GreaterThanOrEqualPredicate;
 import com.redis.om.spring.search.stream.predicates.numeric.GreaterThanPredicate;
 import com.redis.om.spring.search.stream.predicates.numeric.LessThanOrEqualPredicate;
 import com.redis.om.spring.search.stream.predicates.numeric.LessThanPredicate;
+import com.redis.om.spring.search.stream.predicates.numeric.NotEqualPredicate;
```
Comment (chayim) [inline redis-om-spring/src/main/java/com/redis/om/spring/metamodel/indexed/DateField.java:12]: In ULIDIdentifiereGenerator I see that we specifically ensure ULID, ... integer and long.

But here, we only index on DateField, GeoField, and TagField. Can we add Numeric? I recognize we don't (yet) have Vector available here.

File: redis-om-spring/src/main/java/com/redis/om/spring/repository/query/QueryUtils.java [hunk 6]
```diff
@@ -3,19 +3,29 @@
 import java.util.Set;
 
 public class QueryUtils {
-  public static Set<Character> tagEscapeChars = Set.of( //
+  public static final Set<Character> TAG_ESCAPE_CHARS = Set.of( //
```
Comment (chayim) [inline redis-om-spring/src/main/java/com/redis/om/spring/repository/query/QueryUtils.java:6]: IMHO these might generally be useful for JSON as well. Perhaps they belong elsewhere?

File: redis-om-spring/src/main/java/com/redis/om/spring/repository/query/RediSearchQuery.java [hunk 7]
```diff
@@ -266,44 +270,54 @@ public QueryMethod getQueryMethod() {
...
         Pageable pageable = maybePageable.get();
-        query.limit(Long.valueOf(pageable.getOffset()).intValue(), pageable.getPageSize());
-
-        if (pageable.getSort() != null) {
-          for (Order order : pageable.getSort()) {
-            query.setSortBy(order.getProperty(),  order.isAscending());
+        if (!pageable.isUnpaged()) {
+          query.limit(Long.valueOf(pageable.getOffset()).intValue(), pageable.getPageSize());
+  
+          if (pageable.getSort() != null) {
```
Comment (chayim) [inline redis-om-spring/src/main/java/com/redis/om/spring/repository/query/RediSearchQuery.java:289]: What about not paged, but sortable? Or did I misunderstand (not unlikely)

File: redis-om-spring/src/main/java/com/redis/om/spring/repository/query/RediSearchQuery.java [hunk 8]
```diff
@@ -328,15 +342,18 @@ private Object executeQuery(Object[] parameters) {
...
+    }
     AggregationResult aggregationResult = ops.aggregate(aggregation);
 
     // what to return
     Object result = null;
     if (queryMethod.getReturnedObjectType() == AggregationResult.class) {
       result = aggregationResult;
     } else if (queryMethod.isCollectionQuery()) {
-      result = Collections.EMPTY_LIST;
+      result = Collections.EMPTY_LIST; // TODO: WTF Sam-Bodden?
```
Comment (chayim) [inline redis-om-spring/src/main/java/com/redis/om/spring/repository/query/RediSearchQuery.java:356]: At least we should remove the WTF :p

File: redis-om-spring/src/main/java/com/redis/om/spring/convert/BooleanToBytesConverter.java (hunk 1 above)
Comment (bsbodden) [reply redis-om-spring/src/main/java/com/redis/om/spring/convert/BooleanToBytesConverter.java:1]: No exactly, these are Spring Data-specific converters based on the needs for searching. Spring Data provides a pluggable converter API which Spring Data Redis extends, and we further extend it to ensure the conversions for Hashes match the conversions we do for JSON documents. Therefore ensuring consistent serialization/deserialization between Hashes and JSON mapped entities.

File: redis-om-spring/src/main/java/com/redis/om/spring/convert/BooleanToBytesConverter.java (hunk 2 above)
Comment (bsbodden) [reply redis-om-spring/src/main/java/com/redis/om/spring/convert/BooleanToBytesConverter.java:13]: Following the Spring Data Redis lead, we serialize to UTF8 and deserialize from UTF8. We control the serialization charset used in and out of the ORM layer.

File: redis-om-spring/src/main/java/com/redis/om/spring/convert/BytesToBooleanConverter.java (hunk 3 above)
Comment (bsbodden) [reply redis-om-spring/src/main/java/com/redis/om/spring/convert/BytesToBooleanConverter.java:23]: See comment above

File: redis-om-spring/src/main/java/com/redis/om/spring/convert/BytesToUlidConverter.java (hunk 4 above)
Comment (bsbodden) [reply redis-om-spring/src/main/java/com/redis/om/spring/convert/BytesToUlidConverter.java:9]: We can evaluate and swap if needed. Let's create an issue to follow up.

Comment (bsbodden) [reply redis-om-spring/src/main/java/com/redis/om/spring/convert/BytesToUlidConverter.java:9]: See https://github.com/redis/redis-om-spring/issues/87

File: redis-om-spring/src/main/java/com/redis/om/spring/metamodel/indexed/DateField.java (hunk 5 above)
Comment (bsbodden) [reply redis-om-spring/src/main/java/com/redis/om/spring/metamodel/indexed/DateField.java:12]: No, I haven't tackled Vector yet.

File: redis-om-spring/src/main/java/com/redis/om/spring/repository/query/QueryUtils.java (hunk 6 above)
Comment (bsbodden) [reply redis-om-spring/src/main/java/com/redis/om/spring/repository/query/QueryUtils.java:6]: I am using these QueryUtils across both JSON and Hash queries.

File: redis-om-spring/src/main/java/com/redis/om/spring/repository/query/RediSearchQuery.java (hunk 7 above)
Comment (bsbodden) [reply redis-om-spring/src/main/java/com/redis/om/spring/repository/query/RediSearchQuery.java:289]: The is support for that too, the pattern in Spring Data seems to pass a Pageable that has defaults for the paging, e.g. 0 to Collection.size and include the Sort object as needed

File: redis-om-spring/src/main/java/com/redis/om/spring/repository/query/RediSearchQuery.java (hunk 8 above)
Comment (bsbodden) [reply redis-om-spring/src/main/java/com/redis/om/spring/repository/query/RediSearchQuery.java:356]: I like to "keep it real" :-) I got a PR after this one that addresses this... Aggregations support is pretty light right now.

File: redis-om-spring/src/main/java/com/redis/om/spring/convert/BooleanToBytesConverter.java (hunk 1 above)
Comment (bsbodden) [reply redis-om-spring/src/main/java/com/redis/om/spring/convert/BooleanToBytesConverter.java:1]: See https://github.com/spring-projects/spring-data-redis/blob/main/src/main/java/org/springframework/data/redis/core/convert/BinaryConverters.java

File: redis-om-spring/src/main/java/com/redis/om/spring/repository/query/RediSearchQuery.java (hunk 8 above)
Comment (bsbodden) [reply redis-om-spring/src/main/java/com/redis/om/spring/repository/query/RediSearchQuery.java:356]: See https://github.com/redis/redis-om-spring/issues/88

**Review by chayim — APPROVED** [review]

**Pull request merged**